### PR TITLE
MSI Dev App : MSI Helper Service to assist with MSI Integration testing

### PR DESCRIPTION
**Pull request by gladjohn**
Fixes [#3843](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/3843)

**Changes proposed in this request**
MSI Helper Service. 

- Adds functionality to get Environment variables from IMDS, Azure Web App and Azure Function 
- Get MSI Token from IMDS, Azure Web App and Azure Function 
- ReadMe on setup and how it works
- also created a staging slot for service resiliency and added notes to the readme on how to deploy and test using staging slot, before pushing to PROD

**Notes**
- main code is on the controller calling into [MSIHelper ](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/6b9444e56b6e60777857dbbb623137f4932ad78b/tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs)
- you can use the [swagger](https://msalmsiwebapp.azurewebsites.net/swagger/index.html)  to test the app, you need to have lab access to do so

**Testing**
using the test service [swagger](https://msalmsiwebapp.azurewebsites.net/swagger/index.html) 

**Performance impact**
none

**Documentation**
- [ ] All relevant documentation is updated.

### CODE REVIEW

**Comment by neha-bhargava** [comment]
Should we add a readme with this? Do we need to publish this somewhere?

**Comment by gladjohn** [comment]
> Should we add a readme with this? Do we need to publish this somewhere?

Yes a readme will be good to have. I will work on adding one

File: tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+using System.Net.WebSockets;
+using System.Text;
+using System.Text.Json.Serialization;
+using System.Web;
+
+namespace MSIHelperService.Helper
+{
+    internal static class MSIHelper
+    {
+        private static HttpClient s_httpClient = new HttpClient();
```
Comment (bgavrilMS) [inline tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs:24]: nit: consider using IHttpClientFactory from ASP.NET Core https://learn.microsoft.com/en-us/aspnet/core/fundamentals/http-requests?view=aspnetcore-7.0

File: tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+        internal static string? s_functionAppUri = Environment.GetEnvironmentVariable("functionAppUri");
+        internal static string? s_functionAppEnvCode = Environment.GetEnvironmentVariable("functionAppEnvCode");
+        internal static string? s_functionAppMSICode = Environment.GetEnvironmentVariable("functionAppMSICode");
+        internal static string? s_webhookLocation = Environment.GetEnvironmentVariable("webhookLocation");
+        internal static string? s_oMSAdminClientID = Environment.GetEnvironmentVariable("OMSAdminClientID");
+        internal static string? s_oMSAdminClientSecret = Environment.GetEnvironmentVariable("OMSAdminClientSecret");
+
+        //Protected from this API so the getEnvironmentVariable does not return these
+        internal static string[] s_protectedEnvironmentVariables = new string[] {
+            "requestAppID",
```
Comment (bgavrilMS) [inline tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs:45]: Security wise, it's better to have an allow-list rather than a deny list, but this would make the test less valuable.

File: tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+        internal static string? s_requestAppSecret = Environment.GetEnvironmentVariable("requestAppSecret");
+        internal static string? s_functionAppUri = Environment.GetEnvironmentVariable("functionAppUri");
+        internal static string? s_functionAppEnvCode = Environment.GetEnvironmentVariable("functionAppEnvCode");
+        internal static string? s_functionAppMSICode = Environment.GetEnvironmentVariable("functionAppMSICode");
+        internal static string? s_webhookLocation = Environment.GetEnvironmentVariable("webhookLocation");
+        internal static string? s_oMSAdminClientID = Environment.GetEnvironmentVariable("OMSAdminClientID");
+        internal static string? s_oMSAdminClientSecret = Environment.GetEnvironmentVariable("OMSAdminClientSecret");
+
+        //Protected from this API so the getEnvironmentVariable does not return these
+        internal static string[] s_protectedEnvironmentVariables = new string[] {
```
Comment (bgavrilMS) [inline tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs:44]: Pls add "readonly" where possible

File: tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+            //On the Web app we have stored some environment variables, those variables 
+            //hold values that are used to call into several other azure resources
+            //Hide the App ID and App Secret and other variableds in this API
+            //these are used to get token for other azure test resources and stored in the
+            //web apps environment variables
+
+            logger.LogInformation("GetWebAppEnvironmentVariable Function called.");
+
+            return
+                s_protectedEnvironmentVariables.Any(s => s.Contains(variableName, StringComparison.OrdinalIgnoreCase)) ?
```
Comment (bgavrilMS) [inline tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs:99]: Instead of sending NULL, can you send a constant like "REDACTED" or smth like that? That would make the test more useful I think, because having a bad env var value is more useful than not having an env var.

Comment (bgavrilMS) [inline tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs:99]: You should probably also filter the env variables that have "secret" in their name.

File: tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+            return keyValuePairs;
+        }
+
+        /// <summary>
+        /// Gets the Environment Variable from the Azure Function
+        /// </summary>
+        /// <param name="variableName"></param>
+        /// <param name="logger"></param>
+        /// <returns></returns>
+        public static string? GetFunctionAppEnvironmentVariable(
```
Comment (bgavrilMS) [inline tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs:131]: Why don't you make these methods async to avoid Task.Run ?

File: tests/devapps/MSI/HelperService/MSIHelperService/Controllers/GetEnvironmentVariableController.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        /// <param name="variableName">Environment Variable Name Requested for</param>
+        /// <param name="resource">Allowed Values : "webapp", "function", "vm", "azurearc", "servicefabric"</param>
+        /// <returns>
+        /// Returns the environment variable of an azure resource based on the value passed on the query string 
+        /// </returns>
+        [SwaggerResponse(200, "Returns Azure Web App Environment Variables", Type = typeof(string))]
+        [SwaggerResponse(400, "Returns the error object for any validation failures", Type = typeof(string))]
+        [SwaggerResponse(500, "Returns the error object for any Server Errors", Type = typeof(string))]
+        [HttpGet]
+        public string? GetEnvValue(
```
Comment (bgavrilMS) [inline tests/devapps/MSI/HelperService/MSIHelperService/Controllers/GetEnvironmentVariableController.cs:38]: Is this controller still needed?

File: tests/devapps/MSI/HelperService/MSIHelperService/Controllers/GetEnvironmentVariablesController.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+using Microsoft.AspNetCore.Mvc;
+using MSIHelperService.Helper;
+using Swashbuckle.AspNetCore.Annotations;
+
+namespace MSIHelperService.Controllers
+{
+    [ApiController]
+    [Route("[controller]")]
+    [SwaggerTag(description: "Gets All Environment Variables")]
+    public class GetEnvironmentVariablesController : ControllerBase
```
Comment (bgavrilMS) [inline tests/devapps/MSI/HelperService/MSIHelperService/Controllers/GetEnvironmentVariablesController.cs:13]: Why we shouldn't protect this API and instead rely on obfuscating values ? 

https://github.com/AzureAD/microsoft-identity-web/wiki/web-apis

File: tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs (hunk 5 above)
Comment (bgavrilMS) [reply tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs:131]: Generally it's better to have everything async. Hiding async is bad practice, it's inneficient (you spin an extra thread in that Task.Run). In some languages like NodeJS you can't even do sync over async - there's always a way to expose async :)

File: tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+                //return ManagedIdentityResponse object so we can trim the access token
+                return new ContentResult { 
+                    Content = ConvertToJsonString(managedIdentityResponse), 
+                    ContentType = contentEncoding, 
+                    StatusCode = statusCode 
+                };
+            }
+
+            //return errors as is from the MSI Endpoints
+            return new ContentResult {
```
Comment (bgavrilMS) [inline tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs:363]: What about Http exceptions?

File: tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+        }
+
+        /// <summary>
+        /// Gets the MSI Token from the Azure Function App
+        /// </summary>
+        /// <param name="identityHeader"></param>
+        /// <param name="uri"></param>
+        /// <param name="logger"></param>
+        /// <returns></returns>
+        public static async Task<ActionResult?> GetFunctionAppMSIToken(
```
Comment (bgavrilMS) [inline tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs:231]: How is this going to play with MSAL? Because MSAL allows injection of HttpClient, but you are returning a full response here (ManagedIdentityResponse)

File: tests/devapps/MSI/HelperService/MSIHelperService/Controllers/GetMSITokenController.cs [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        /// <param name="azureResource">Resource for which you need the MSI Token</param>
+        /// <returns>
+        /// Returns the MSI token for an azure resource
+        /// </returns>
+        [ProducesResponseType(typeof(ManagedIdentityResponse), (int)MSIHelper.HTTPErrorResponseCode.Status200OK)]
+        [HttpGet]
+        [SwaggerResponse(200, "Returns an Azure Resource MSI Token Response", Type = typeof(ManagedIdentityResponse))]
+        [SwaggerResponse(400, "Returns the error object for any validation failures", Type = typeof(ManagedIdentityResponse))]
+        [SwaggerResponse(500, "Returns the error object for any Server Errors", Type = typeof(ManagedIdentityResponse))]
+        public async Task<ActionResult?> GetRemoteHttpResponse(
```
Comment (bgavrilMS) [inline tests/devapps/MSI/HelperService/MSIHelperService/Controllers/GetMSITokenController.cs:40]: I am not sure how MSAL tests would be able to use this, because MSAL tests expect HttpResponseMessage and not `ManagedIdentityResponse`. 

Not blocking but would like to see a test.

Comment (bgavrilMS) [reply tests/devapps/MSI/HelperService/MSIHelperService/Controllers/GetMSITokenController.cs:40]: i.e. in my mind the interface would be more general purpose

`GetRemoteHttpResponse(string uri, HttpMethod method, string body_for_post, Dictionary<string, string> requestHeaders)`

File: tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs (hunk 2 above)
Comment (neha-bhargava) [reply tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs:45]: I think we can have all the environment variables expected by managed identity flow in the allow list.

File: tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs (hunk 1 above)
Comment (gladjohn) [reply tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs]: thanks for sharing this, this is very effective on a web app

File: tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs (hunk 4 above)
Comment (gladjohn) [reply tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs]: it's protected

File: tests/devapps/MSI/HelperService/MSIHelperService/Controllers/GetEnvironmentVariableController.cs (hunk 6 above)
Comment (gladjohn) [reply tests/devapps/MSI/HelperService/MSIHelperService/Controllers/GetEnvironmentVariableController.cs]: deleted

File: tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs (hunk 8 above)
Comment (gladjohn) [reply tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs:308]: this should return all error messages and exceptions from MSI 

**Sample validation message from MSI :** 

`curl -X 'GET' \
  'https://service.msidlab.com/GetMSIToken?uri=http%3A%2F%2F127.0.0.1%3A41292%2Fmsi%2Ftoken%2F&azureResource=webapp' \
  -H 'accept: text/plain' \
  -H 'X-IDENTITY-HEADER: 69C62B109AAF4EB7B01061197C14F550'`

`"{\"error\":{\"code\":\"ApiVersionUnspecified\",\"message\":\"An API version is required, but was not specified.\",\"innerError\":null}}"`

**Wrong API Version :** 

`curl -X 'GET' \
  'https://service.msidlab.com/GetMSIToken?uri=http%3A%2F%2F127.0.0.1%3A41292%2Fmsi%2Ftoken%2F%3Fapi-version%3D2019-01-01%26resource%3Dhttps%3A%2F%2Fmanagement.azure.com&azureResource=webapp' \
  -H 'accept: text/plain' \
  -H 'X-IDENTITY-HEADER: 69C62B109AAF4EB7B01061197C14F550'`

`"{\"error\":{\"code\":\"UnsupportedApiVersion\",\"message\":\"The HTTP resource that matches the request URI 'http://127.0.0.1:41292/msi/token' does not support the API version '2019-01-01'.\",\"innerError\":null}}"`

**Exceptions**

`curl -X 'GET' \
  'https://service.msidlab.com/GetMSIToken?uri=http%3A%2F%2F127.0.0.1%3A41292%2Fmsi%2Ftoken%2F%3Fapi-version%3D2019-08-01%26resource%3Dhttps%3A%2F%2Fmanagement.azure.com&azureResource=function' \
  -H 'accept: text/plain' \
  -H 'X-IDENTITY-HEADER: 69C62B109AAFEB7B01061197C14F550'`

`"An attempt was made to access a socket in a way forbidden by its access permissions. (127.0.0.1:41292) \n\nAn attempt was made to access a socket in a way forbidden by its access permissions."
`

Comment (gladjohn) [reply tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs:308]: With the error codes

image

File: tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs (hunk 9 above)
Comment (gladjohn) [reply tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs:166]: Added the integration test PR, please take a look at how we consume this MSIHelper service within MSAL

**Comment by gladjohn** [comment]
> > Should we add a readme with this? Do we need to publish this somewhere?
> 
> Yes a readme will be good to have. I will work on adding one

added

File: tests/devapps/MSI/HelperService/MSIHelperService/ManagedIdentityResponse.cs [hunk 11]
```diff
@@ -6,27 +6,45 @@
 
 namespace MSIHelperService
 {
+    /// <summary>
+    /// ManagedIdentityResponse
+    /// </summary>
     [JsonObject]
     public class ManagedIdentityResponse
```
Comment (bgavrilMS) [inline tests/devapps/MSI/HelperService/MSIHelperService/ManagedIdentityResponse.cs:13]: I think Neha added some error fields to this one recently.

File: tests/devapps/MSI/HelperService/MSIHelperService/AzureFunction/GetEnvironmentVariables.cs [hunk 12]
```diff
@@ -0,0 +1,60 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+//#r "Newtonsoft.Json" //uncomment this on the function app
```
Comment (neha-bhargava) [inline tests/devapps/MSI/HelperService/MSIHelperService/AzureFunction/GetEnvironmentVariables.cs:4]: Isn't this for the function app?

File: tests/devapps/MSI/HelperService/MSIHelperService/ManagedIdentityResponse.cs (hunk 11 above)
Comment (neha-bhargava) [reply tests/devapps/MSI/HelperService/MSIHelperService/ManagedIdentityResponse.cs:13]: There is a ManagedIdentityErrorResponse for error fields

File: tests/devapps/MSI/HelperService/MSIHelperService/AzureFunction/GetEnvironmentVariables.cs (hunk 12 above)
Comment (gladjohn) [reply tests/devapps/MSI/HelperService/MSIHelperService/AzureFunction/GetEnvironmentVariables.cs:4]: Yes, this code is deployed on the function app. commented that function app line so the project can build

File: tests/devapps/MSI/HelperService/MSIHelperService/ManagedIdentityResponse.cs (hunk 11 above)
Comment (gladjohn) [reply tests/devapps/MSI/HelperService/MSIHelperService/ManagedIdentityResponse.cs:13]: The only reason we use `ManagedIdentityResponse` is to trim the access token, else we can send the response as is so MSAL can handle it. But for error response there is nothing to filter so we do not transform and send it to MSAL as is

File: tests/devapps/MSI/HelperService/MSIHelperService/AzureFunction/GetEnvironmentVariables.cs [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+                _envVariables.Add("IDENTITY_API_VERSION", Environment.GetEnvironmentVariable("IDENTITY_API_VERSION"));
+
+                log.LogInformation("Returning All Environment Variables");
+                return new OkObjectResult(_envVariables);
+            }
+            else
+            {
+                string? responseMessage = Environment.GetEnvironmentVariable(name);
+#pragma warning disable CA2254 // Template should be a static expression
+                log.LogInformation($"Returning Environment Variables Based on the variable name : { name } ");
```
Comment (neha-bhargava) [inline tests/devapps/MSI/HelperService/MSIHelperService/AzureFunction/GetEnvironmentVariables.cs:54]:
```suggestion
                log.LogInformation($"Returning Environment Variable Based on the variable name : { name } ");
```

File: tests/devapps/MSI/HelperService/MSIHelperService/AzureFunction/GetEnvironmentVariables.cs [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+            string name = req.Query["variableName"];
+
+            string requestBody = await new StreamReader(req.Body).ReadToEndAsync().ConfigureAwait(false);
+            dynamic? data = JsonConvert.DeserializeObject(requestBody);
+#pragma warning disable CS8600 // Converting null literal or possible null value to non-nullable type.
+            name = name ?? data?.variableName;
+#pragma warning restore CS8600 // Converting null literal or possible null value to non-nullable type.
+
+#pragma warning disable CA2254 // Template should be a static expression
+            log.LogInformation($"Querysting value for  variableName is : { name } ");
```
Comment (neha-bhargava) [inline tests/devapps/MSI/HelperService/MSIHelperService/AzureFunction/GetEnvironmentVariables.cs:37]:
```suggestion
            log.LogInformation($"Querystring value for  variableName is : { name } ");
```

File: tests/devapps/MSI/HelperService/MSIHelperService/.config/dotnet-tools.json [hunk 15]
```diff
@@ -0,0 +1,12 @@
+{
+  "version": 1,
+  "isRoot": true,
+  "tools": {
+    "dotnet-ef": {
```
Comment (bgavrilMS) [inline tests/devapps/MSI/HelperService/MSIHelperService/.config/dotnet-tools.json:5]: What does this json do? Will it install Entity Framework?

File: tests/devapps/MSI/HelperService/MSIHelperService/MSIHelperService.csproj [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <Nullable>enable</Nullable>
+    <ImplicitUsings>enable</ImplicitUsings>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.ApplicationInsights.AspNetCore" Version="2.21.0" />
+    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
```
Comment (bgavrilMS) [inline tests/devapps/MSI/HelperService/MSIHelperService/MSIHelperService.csproj:11]: No more newtonsoft pls.

File: tests/devapps/MSI/HelperService/MSIHelperService/AzureFunction/GetEnvironmentVariables.cs [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+
+            string name = req.Query["variableName"];
+
+            string requestBody = await new StreamReader(req.Body).ReadToEndAsync().ConfigureAwait(false);
+            dynamic? data = JsonConvert.DeserializeObject(requestBody);
+#pragma warning disable CS8600 // Converting null literal or possible null value to non-nullable type.
+            name = name ?? data?.variableName;
+#pragma warning restore CS8600 // Converting null literal or possible null value to non-nullable type.
+
+#pragma warning disable CA2254 // Template should be a static expression
```
Comment (bgavrilMS) [inline tests/devapps/MSI/HelperService/MSIHelperService/AzureFunction/GetEnvironmentVariables.cs:36]: Please, let's not disable any warnings. Instead, work to fix the root cause identified.

**Comment by bgavrilMS** [comment]
Needs a code walkthrough !

File: tests/devapps/MSI/HelperService/MSIHelperService/AzureFunction/GetEnvironmentVariables.cs (hunk 14 above)
Comment (gladjohn) [reply tests/devapps/MSI/HelperService/MSIHelperService/AzureFunction/GetEnvironmentVariables.cs]: fixed

File: tests/devapps/MSI/HelperService/MSIHelperService/AzureFunction/GetEnvironmentVariables.cs [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+/// Name space is not required while updating this code in the function app
+/// </summary>
+namespace MSIHelperService.AzureFunction
+{
+    /// <summary>
+    /// Gets the environment variables from the function app 
+    /// </summary>
+    public class GetEnvironmentVariables
+    {
+        private static Dictionary<string, string> _envVariables => new Dictionary<string, string>();
```
[the remark this answers is not in the document]
Comment (gladjohn) [reply tests/devapps/MSI/HelperService/MSIHelperService/AzureFunction/GetEnvironmentVariables.cs]: this is used in the Azure Function. Just having the code in the "AzureFunction" folder so we can add it to the the repo

File: tests/devapps/MSI/HelperService/MSIHelperService/ManagedIdentityResponse.cs [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+using System;
+using Newtonsoft.Json;
+
+namespace MSIHelperService
+{
+    /// <summary>
+    /// ManagedIdentityResponse
+    /// </summary>
+    [JsonObject]
+    public class ManagedIdentityResponse
```
[the remark this answers is not in the document]
Comment (gladjohn) [reply tests/devapps/MSI/HelperService/MSIHelperService/ManagedIdentityResponse.cs]: removed not needed

File: tests/devapps/MSI/HelperService/MSIHelperService/MSIHelperService.csproj (hunk 16 above)
Comment (gladjohn) [reply tests/devapps/MSI/HelperService/MSIHelperService/MSIHelperService.csproj]: not needed

File: tests/devapps/MSI/HelperService/MSIHelperService/readme.md [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+manager: bogavril
+
+#Customer intent: As a developer I would like to test Managed Service Identity on my dev box and run integration tests. This document provides the test setup on how this can be achieved.
+---
+
+# Testing managed identities on Azure resources for MSAL .Net and other Auth SDKs
+
+A common challenge for developers testing Managed Identities is to deploy their code to a Managed Identity Service, so they can send requests to the Managed Identity endpoints. While this may be feasible for an end-to-end testing scenario, but it could be challenging to run integration tests on CI pipelines or on a local dev box.
+
+To eliminate this problem, MSAL .Net team has built a MSI Helper service that exposes several managed identity sources through a web api using which you can access the identity endpoints from your tests and acquire Managed Identity token responses. The API accepts the same query params of a managed identity endpoint, this will allow your test(s) to call the api just like how you would make a http request to the managed identity endpoint while your code runs inside of a managed identity source.
```
[the remark this answers is not in the document]
Comment (gladjohn) [reply tests/devapps/MSI/HelperService/MSIHelperService/readme.md]: unfortunately all these have been outdated and I am not able to commit directly sorry

File: tests/devapps/MSI/HelperService/MSIHelperService/.config/dotnet-tools.json (hunk 15 above)
Comment (gladjohn) [reply tests/devapps/MSI/HelperService/MSIHelperService/.config/dotnet-tools.json]: yes it came as part of the weather sample. removed it

File: tests/devapps/Managed Identity apps/MSIHelperService/Helper/MSIHelper.cs [hunk 21]
```diff
@@ -0,0 +1,518 @@
...
+        //Enum for HTTP Error Response Codes
+        internal enum HTTPErrorResponseCode
+        {
+            Status200OK = StatusCodes.Status200OK,
+            Status201Created = StatusCodes.Status201Created,
+            Status400BadRequest = StatusCodes.Status400BadRequest,
+            Status404NotFound = StatusCodes.Status404NotFound,
+            Status500InternalServerError = StatusCodes.Status500InternalServerError,
+            Status503ServiceUnavailable = StatusCodes.Status503ServiceUnavailable
+        }
```
Comment (pmaytak) [inline tests/devapps/Managed Identity apps/MSIHelperService/Helper/MSIHelper.cs:60]: nit: remove? unused?

File: tests/devapps/Managed Identity apps/MSIHelperService/AzureFunction/GetEnvironmentVariables.cs [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+
+            log.LogInformation("Querystring value for variableName is : ", name);
+
+            if (string.IsNullOrEmpty(name))
+            {
+                Dictionary<string, string?> _envVariables = new Dictionary<string, string?>();
+                _envVariables.Add("IDENTITY_HEADER", Environment.GetEnvironmentVariable("IDENTITY_HEADER"));
+                _envVariables.Add("IDENTITY_ENDPOINT", Environment.GetEnvironmentVariable("IDENTITY_ENDPOINT"));
+                _envVariables.Add("IDENTITY_API_VERSION", Environment.GetEnvironmentVariable("IDENTITY_API_VERSION"));
+                await Task.Delay(1).ConfigureAwait(true);
```
Comment (pmaytak) [inline tests/devapps/Managed Identity apps/MSIHelperService/AzureFunction/GetEnvironmentVariables.cs:39]: Why is this needed?

**Review by pmaytak — APPROVED** [review]

File: tests/devapps/Managed Identity apps/MSIHelperService/AzureFunction/GetToken.cs [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+using System.Net;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Primitives;
+
+namespace MSIHelperService.AzureFunction
+{
+    /// <summary>
+    /// Gets the function app token
+    /// </summary>
+    public class GetToken
```
Comment (neha-bhargava) [inline tests/devapps/Managed Identity apps/MSIHelperService/AzureFunction/GetToken.cs:17]: nit: AquireToken?
